packages/mermaid/src/diagrams/xychart/parser/xychart.jison.spec.ts

```diff
@@ -0,0 +1,367 @@
+// @ts-ignore: TODO Fix ts errors
+import { parser } from './xychart.jison';
+import { Mock, vi } from 'vitest';
+
+const parserFnConstructor = (str: string) => {
+  return () => {
+    parser.parse(str);
+  };
+};
+
+const mockDB: Record<string, Mock<any, any>> = {
+  parseDirective: vi.fn(),
+  setOrientation: vi.fn(),
+  setDiagramTitle: vi.fn(),
+  setXAxisTitle: vi.fn(),
+  setXAxisRangeData: vi.fn(),
+  setXAxisBand: vi.fn(),
+  setYAxisTitle: vi.fn(),
+  setYAxisRangeData: vi.fn(),
+  setLineData: vi.fn(),
+  setBarData: vi.fn(),
+};
+
+function clearMocks() {
+  for (const key in mockDB) {
+    mockDB[key].mockRestore();
+  }
+}
+
+describe('Testing xychart jison file', () => {
+  beforeEach(() => {
+    parser.yy = mockDB;
+    clearMocks();
+  });
+
+  it('should throw error if xychart-beta text is not there', () => {
+    const str = 'xychart-beta-1';
+    expect(parserFnConstructor(str)).toThrow();
+  });
+
+  it('should not throw error if only xychart is there', () => {
+    const str = 'xychart-beta';
+    expect(parserFnConstructor(str)).not.toThrow();
+  });
+
+  it('parse title of the chart', () => {
+    const str = 'xychart-beta \n title This is a title';
+    expect(parserFnConstructor(str)).not.toThrow();
+    expect(mockDB.setDiagramTitle).toHaveBeenCalledWith('This is a title');
+  });
+
+  it('should be able to parse directive', () => {
+    const str =
+      '%%{init: {"xychart": {"chartWidth": 600, "chartHeight": 600} } }%%  \n xychart-beta';
+    expect(parserFnConstructor(str)).not.toThrow();
+    expect(mockDB.parseDirective.mock.calls[0]).toEqual(['%%{', 'open_directive']);
+    expect(mockDB.parseDirective.mock.calls[1]).toEqual(['init', 'type_directive']);
+    expect(mockDB.parseDirective.mock.calls[2]).toEqual([
+      '{"xychart": {"chartWidth": 600, "chartHeight": 600} }',
+      'arg_directive',
+    ]);
+    expect(mockDB.parseDirective.mock.calls[3]).toEqual(['}%%', 'close_directive', 'xychart']);
+  });
+
+  it('parse chart orientation', () => {
+    let str = 'xychart-beta vertical';
+    expect(parserFnConstructor(str)).not.toThrow();
+    expect(mockDB.setOrientation).toHaveBeenCalledWith('vertical');
+
+    clearMocks();
+
+    str = 'xychart-beta        horizontal        ';
+    expect(parserFnConstructor(str)).not.toThrow();
+    expect(mockDB.setOrientation).toHaveBeenCalledWith('horizontal');
+
+    str = 'xychart-beta abc';
+    expect(parserFnConstructor(str)).toThrow();
+  });
+
+  it('parse x-axis', () => {
+    let str = 'xychart-beta \nx-axis xAxisName\n';
+    expect(parserFnConstructor(str)).not.toThrow();
+    expect(mockDB.setXAxisTitle).toHaveBeenCalledWith({
+      text: 'xAxisName',
+      type: 'text',
+    });
+
+    clearMocks();
+
+    str = 'xychart-beta \nx-axis        xAxisName     \n';
+    expect(parserFnConstructor(str)).not.toThrow();
+    expect(mockDB.setXAxisTitle).toHaveBeenCalledWith({
+      text: 'xAxisName',
+      type: 'text',
+    });
+
+    clearMocks();
```

Comment:
I see some repetitive `clearMocks()` calls and at the same time you have set up to `clearMocks` before each example, so these chunks should be probably split into separate `it` calls.

Reply:
Refactored the test cases

Reply:
Nice!